PhysicsTools/BPHNano/plugins/V0ReBuilder.cc

```diff
@@ -0,0 +1,206 @@
+#include "TrackingTools/Records/interface/TransientTrackRecord.h"
+#include "TrackingTools/TransientTrack/interface/TransientTrack.h"
+#include "TrackingTools/TransientTrack/interface/TransientTrackBuilder.h"
+#include "helper.h"
+
+class V0ReBuilder : public edm::global::EDProducer<> {
+  // perhaps we need better structure here (begin run etc)
+public:
+  typedef std::vector<reco::TransientTrack> TransientTrackCollection;
+  typedef std::vector<reco::VertexCompositePtrCandidate> V0Collection;
+
+  explicit V0ReBuilder(const edm::ParameterSet &cfg)
+      : theB_(esConsumes(edm::ESInputTag{"", "TransientTrackBuilder"})),
+        trk_selection_{cfg.getParameter<std::string>("trkSelection")},
+        pre_vtx_selection_{cfg.getParameter<std::string>("V0Selection")},
+        post_vtx_selection_{cfg.getParameter<std::string>("postVtxSelection")},
+        v0s_{consumes<V0Collection>(cfg.getParameter<edm::InputTag>("V0s"))},
+        beamspot_{consumes<reco::BeamSpot>(cfg.getParameter<edm::InputTag>("beamSpot"))},
+        track_match_{consumes<edm::Association<pat::CompositeCandidateCollection>>(
+            cfg.getParameter<edm::InputTag>("track_match"))},
+        isLambda_{cfg.getParameter<bool>("isLambda")} {
+    produces<pat::CompositeCandidateCollection>("SelectedV0Collection");
+    produces<TransientTrackCollection>("SelectedV0TransientCollection");
+  }
+
+  ~V0ReBuilder() override {}
+
+  void produce(edm::StreamID, edm::Event &, const edm::EventSetup &) const override;
+
+private:
+  const edm::ESGetToken<TransientTrackBuilder, TransientTrackRecord> theB_;
+  const StringCutObjectSelector<pat::PackedCandidate> trk_selection_;
+  const StringCutObjectSelector<reco::VertexCompositePtrCandidate> pre_vtx_selection_;
+  const StringCutObjectSelector<pat::CompositeCandidate> post_vtx_selection_;
+  const edm::EDGetTokenT<V0Collection> v0s_;
+  const edm::EDGetTokenT<reco::BeamSpot> beamspot_;
+  const edm::EDGetTokenT<edm::Association<pat::CompositeCandidateCollection>> track_match_;
+  const bool isLambda_;
+};
+
+void V0ReBuilder::produce(edm::StreamID, edm::Event &evt, edm::EventSetup const &iSetup) const {
+  // input
+  auto const theB = &iSetup.getData(theB_);
+  edm::Handle<V0Collection> V0s;
+  evt.getByToken(v0s_, V0s);
+  edm::Handle<reco::BeamSpot> beamspot;
+  evt.getByToken(beamspot_, beamspot);
+
+  auto &track_match = evt.get(track_match_);
+
+  // output
+  std::unique_ptr<pat::CompositeCandidateCollection> ret_val(new pat::CompositeCandidateCollection());
+  std::unique_ptr<TransientTrackCollection> trans_out(new TransientTrackCollection);
+
+  size_t v0_idx = 0;
+  for (reco::VertexCompositePtrCandidateCollection::const_iterator v0 = V0s->begin(); v0 != V0s->end(); v0++) {
+    reco::VertexCompositePtrCandidate V0 = V0s->at(v0_idx);
+    v0_idx++;
+
+    // selection on V0s
+    if (v0->numberOfDaughters() != 2)
+      continue;
+    if (!pre_vtx_selection_(V0))
+      continue;
```

Comment:
Why not using:

```suggestion
    if (!pre_vtx_selection_(*v0))
```

Then you can eliminate `V0` and `v0_idx` completely.